baybe/utils/basic.py

```diff
@@ -181,3 +182,37 @@ def find_subclass(base: type, name_or_abbr: str, /):
             f"The class name or abbreviation '{name_or_abbr}' does not refer to any "
             f"of the subclasses of '{base.__name__}'."
         )
+
+
+def register_hook(target: Callable, hook: Callable) -> Callable:
+    """Register a callable for the given target method.
+
+    Args:
+        target: The method to be changed.
+        hook: The callable to be registered in the given method.
+
+    Returns:
+        A wrapped callable that replaces the original target method.
+
+    Raises:
+        TypeError: If the signature of the callable does not match the signature of the
+            target method.
+    """
+    for p1, p2 in zip(
+        inspect.signature(target, eval_str=True).parameters.values(),
+        inspect.signature(hook, eval_str=True).parameters.values(),
+    ):
+        if p1.name != p2.name or p1.annotation != p2.annotation:
+            print(p2.annotation)
+            if p2.annotation is not inspect.Parameter.empty:
+                raise TypeError(
+                    f"The signature of '{hook.__name__}' does not match the "
+                    f"signature of '{target.__name__}'."
+                )
+
+    @functools.wraps(target)
+    def wraps(*args, **kwargs):
```

Comment:
im never quite sure but since this is supposed to overwrite methdos... do we need `self` here? Do we need a distinction between methods and non-methods?
Theres also a potentially interesting type for methods called `MethodType` to be imported from `types` (not `typing`)

Reply:
I guess I am wondering whether self is in any sense special and needs to be explicitly written or if tis also contained in `args` ?

Reply:
Well, `self` is the first arg that should be included in the hook signature to match the target's signature (which is always the case for class methods) so yes it is included in the args.

Reply:
See also [here](https://github.com/emdgroup/baybe/pull/247#discussion_r1631158559)